### PR TITLE
US121367 convert sirenComponentFactory to sirenObservableFactory and test

### PR DESCRIPTION
**Pull request by tenbradley**
Work for [US121367](https://rally1.rallydev.com/#/357252966636d/custom/367300408400?detail=%2Fuserstory%2F441183081456).

- investigate cleaner ways to connect to HypermediaState.
- Rename basicInfo to definedProperties.
- Unit & Contract Test and test more!

### CODE REVIEW

**Comment by tenbradley** [comment]
At this point, I think the unit testing for sirenObservableFactory is about done.

**Comment by tenbradley** [comment]
Note: tests are note made for siren{Action,Link,SubEntity,SubEntities} as they need updated from stories first

File: test/observables/observable.test.js [hunk 1]
```diff
@@ -0,0 +1,204 @@
+import { assert }  from '@open-wc/testing';
```
Comment (zinabat) [inline test/observables/observable.test.js:1]: Let's keep things consistent. The folder for both tests and source code should either be `observable/` or `observables/`

File: test/observables/sirenClasses.test.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import { Component } from '../../state/observable/Common.js';
+import { SirenClasses } from '../../state/observable/SirenClasses.js';
+
+describe('call sirenClasses methods', () => {
+	it('call constructor for sirenClasses', () => {
+		const obj = new SirenClasses();
+		assert(!obj.value);
+	});
+
+	it('value of SirenClasses is set', () => {
```
Comment (zinabat) [inline test/observables/sirenClasses.test.js:11]: I don't understand the purpose of this test?

**Review by zinabat — COMMENTED** [review]
I'd like @AKobets to look at this more in depth to make sure our test cases have good coverage, but it's looking good.

I think we did end up deciding on a module instead of a class after all

File: test/observables/sirenClasses.test.js (hunk 2 above)
Comment (tenbradley) [reply test/observables/sirenClasses.test.js:11]: I'll go through my tests later and clean them up. As I said, a large portion of them was me learning Javascript objects.

File: test/observables/sirenSubEntity.test.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// import { assert }  from '@open-wc/testing';
```
Comment (AKobets) [inline test/observables/sirenSubEntity.test.js:1]: are you going to implement tests here ?

File: test/observables/sirenSubEntities.test.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// import { assert }  from '@open-wc/testing';
```
Comment (AKobets) [inline test/observables/sirenSubEntities.test.js:1]: are you going to implement tests here ?

File: test/observables/sirenLink.test.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// import { assert }  from '@open-wc/testing';
```
Comment (AKobets) [inline test/observables/sirenLink.test.js:1]: are you going to implement tests here ?

File: test/observables/sirenAction.test.js [hunk 6]
```diff
@@ -0,0 +1,3 @@
+//import { assert }  from '@open-wc/testing';
```
Comment (AKobets) [inline test/observables/sirenAction.test.js:1]: are you going to implement tests here ?

File: test/observable/sirenSubEntity.test.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+// import { assert }  from '@open-wc/testing';
```
Comment (AKobets) [inline test/observable/sirenSubEntity.test.js:1]: are you going to implement tests here ?

File: test/observable/sirenSubEntities.test.js [hunk 8]
```diff
@@ -0,0 +1,3 @@
+// import { assert }  from '@open-wc/testing';
```
Comment (AKobets) [inline test/observable/sirenSubEntities.test.js:1]: are you going to implement tests here ?

File: test/observable/sirenLink.test.js [hunk 9]
```diff
@@ -0,0 +1,3 @@
+// import { assert }  from '@open-wc/testing';
```
Comment (AKobets) [inline test/observable/sirenLink.test.js:1]: are you going to implement tests here ?

File: test/observable/sirenAction.test.js [hunk 10]
```diff
@@ -0,0 +1,3 @@
+//import { assert }  from '@open-wc/testing';
```
Comment (AKobets) [inline test/observable/sirenAction.test.js:1]: are you going to implement tests here ?

File: test/observables/sirenSubEntity.test.js (hunk 3 above)
Comment (tenbradley) [reply test/observables/sirenSubEntity.test.js:1]: Yes. But waiting for some other code to be refactored in other PRs that are causing a circular dependency for some of the classes. Left the files here so I'd know to come back.

File: test/observables/sirenClasses.test.js (hunk 2 above)
Comment (tenbradley) [reply test/observables/sirenClasses.test.js:11]: https://github.com/BrightspaceHypermediaComponents/foundation-engine/pull/14#issuecomment-718216497

File: test/observables/sirenSubEntity.test.js (hunk 3 above)
Comment (tenbradley) [reply test/observables/sirenSubEntity.test.js:1]: Documented needing to wait here: https://github.com/BrightspaceHypermediaComponents/foundation-engine/pull/14#issuecomment-718216497

File: test/siren-observables.test.js [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+		const basicInfo = sirenObservableFactory(propertyInfo);
+		assert(basicInfo instanceof SirenSubEntity);
+	});
+
+	it('ObservableFactory fails to create class of type that does not exist', () => {
+		const propertyInfo = {
+			type: 8
+		};
+		try {
+			sirenObservableFactory(propertyInfo);
```
Comment (AKobets) [inline test/siren-observables.test.js:76]: could be replaced with expect to throw error.

File: test/siren-observables.test.js [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+		}
+	});
+});
+
+describe('creating sirenDefinedProperty incomplete objects', () => {
+
+	it('property is not created from epmty object', () => {
+		const obj = {};
+		const res = sirenDefinedProperty(obj, null);
+		assert(!res);
```
Comment (AKobets) [inline test/siren-observables.test.js:97]: when assertion has also explanatory message it helps reading/debugging test failures.

**Review by m6jones — APPROVED** [review]

**Pull request merged**